experimental/Examples/NFT.idr

```diff
@@ -0,0 +1,230 @@
+total approve : Address -> TokenId -> Either Error Storage
+approve address token =
+  case getApproved token of
+    Left e => Left e
+    Right approvedAdd =>
+      let isApproved = currentCaller == (maybe (owner storage) id approvedAdd)
+          owner = ownerOfToken token in
+          case currentCaller == owner || isApproved of
+            False => Left FailedToAuthenticate
+            True =>
+              Right
+                (record
+                   {tokens =
+                     insert
+                     token
+                     (MkToken
+                       owner
+                       (Just address)
+                     )
+                     currentTok
+                   } storage
+                 )
+
+||| newOp function to update the operator list.
+total newOp : Address -> Bool -> SortedMap Address Bool
+newOp operator isSet =
+  case lookup currentCaller currentAcc of
+    Nothing => insert operator isSet empty
+    Just op => insert operator isSet (opApprovals op)
+
+||| setApprovalForAll [standard function]
+||| let the owner enable or disable an operator.
+||| The operator can manage all NFTs of the owner.
+total setApprovalForAll : (operator : Address) -> Bool -> Storage
+setApprovalForAll operator isSet =
+    record
+      {accounts =
+        insert
+        currentCaller
+        (MkAccount
+          (balanceOf currentCaller)
+          (newOp operator isSet)
+        )
+        currentAcc
+      } storage
+
+||| isApprovedForAll [standard function]
+||| returns whether an address is an authorized operator for another address.
+||| @owner the address that owns the NFTs.
+||| @operator the address that acts on behalf of the owner.
+total isApprovedForAll : (owner : Address) -> (operator : Address) -> Bool
+isApprovedForAll owner operator =
+  case lookup owner currentAcc of
+    Nothing => False
+    Just op =>
+      case lookup operator (opApprovals op) of
+        Nothing => False
+        Just bool => bool
+
+||| modifyAccBal function to modify the account balance.
+total modifyAccBal :
+Address -> (Nat -> Nat) -> SortedMap Address Account -> SortedMap Address Account
+modifyAccBal add f acc =
+  let addAcc = getAccount add acc in
+    insert add (record {ownedTokens $= f} addAcc) acc
+
+||| transfer [standard function]
+||| transfers the ownership of a NFT from one address to another.
+||| @from the address the tokens to be transferred from
+||| @dest the address the tokens to be transferred to
+||| @token the TokenId of the NFT to be transferred
+total transfer :
+(from : Address) -> (dest : Address) -> (token : TokenId) -> Either Error Storage
+transfer from dest token =
+  case ownerOf token of
+    Left e => Left e
+    Right add =>
+      case add == from of
+        False => Left NotOwnedByFromAddress
+        True =>
+          case currentCaller == from ||
+               currentCaller == maybe (owner storage) id (approvedAdd token) ||
```

Comment:
Why are we using the global `storage` here? Shouldn't we be using `storage` as a parameter somehow?

Reply:
This is following the standard that doesn't seem to take `storage` as an input for those functions. The global `storage` here is a fake current storage. I can change it all to taking a `storage` though.

Reply:
Yeah, I think all the functions should take a `storage` parameter

It's probably just an implicit contract `storage` in the Ethereum standard, it's the equivalent to a parameter in our case